python/src/uagents/experimental/dialogues/__init__.py

```diff
@@ -0,0 +1,467 @@
+"""Dialogue class aka. blueprint for protocols"""
+import functools
+import graphlib
+from datetime import datetime, timedelta
+from typing import Any, Awaitable, Callable, Optional, Type
+from uuid import UUID
+
+from uagents import Context, Model, Protocol
+from uagents.storage import KeyValueStore
+
+DEFAULT_SESSION_TIMEOUT = 10
+
+JsonStr = str
+
+MessageCallback = Callable[["Context", str, Any], Awaitable[None]]
+
+
+class Node:
+    """
+    A node represents a state in the dialogue.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        starter: bool = False,
+    ) -> None:
+        self.name = name
+        self.description = description
+        self.starter = starter
+
+
+class Edge:
+    """
+    An edge represents a transition between two states in the dialogue.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        parent: Node,  # tail
+        child: Node,  # head
+        model: Type[Model] = None,
+        func: MessageCallback = lambda *args, **kwargs: None,
+    ) -> None:
+        self.name = name
+        self.description = description
+        self.parent = parent
+        self.child = child
+        self._model = model
+        self._func = func
+
+    @property
+    def model(self) -> Type[Model]:
+        """The message model type that is associated with the edge."""
+        return self._model
+
+    @model.setter
+    def model(self, model: Type[Model]) -> None:
+        self._model = model
+
+    @property
+    def func(self) -> MessageCallback:
+        """The message handler that is associated with the edge."""
+        return self._func
+
+    @func.setter
+    def func(self, func: MessageCallback) -> None:
+        self._func = func
+
+
+class Dialogue(Protocol):
+    """
+    A Dialogue is the local representation of the dialogue.
+
+    - This should be the local representation of the dialogue.
+    - Each participant will have its own instance of this class per dialogue.
+    - A storage will contain all the dialogues that took place, which may be
+      automatically deleted after a certain amount of time.
+    - is meant to simplify the handling of individual messages
+    """
+
+    def __init__(
+        self,
+        name: str,  # mandatory, due to storage naming
+        version: Optional[str] = None,
+        nodes: list[Node] | None = None,
```

Comment:
To maintain support for Python 3.8 and 3.9, we need to use the older notion, importing from the `typing` library.
```suggestion
        nodes: List[Node] | None = None,
```

Reply:
OK, changed